docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,233 @@
+  `postStart` hook, each running pod incurs an overhead for the container runtime and the kubelet;
+- on the control-plane, a higher number of DaemonSets and Pods creates more events for kube-controller-manager, which
+  trigger more reconciliations and status updates.
+
+## Goals
+
+1. Improve the reliability of module loading and unloading
+2. Improve operator upgrade experience
+3. Reduce the overall KMM footprint
+
+## Non-goals
+
+1. Add further control on the deployment of kmods on nodes
+
+## Design
+
+### KMM: worker Pods pulling kmod images
+
+This proposal moves from ModuleLoader DaemonSets to standalone, one-shot Pods.
+KMM is now composed of two components:
+- the operator;
+- the worker Pods.
+
+```
+               ┌───────┐
+               │       │
+               │ Nodes ◄──┐
+               │       │  │
+               └───────┘  │ watches
+                          │
+             ┌─────────┐  │
+             │         │  │
+             │ Modules │  │     ┌───────────────────────┐
+             │         │  │     │                       │
+             └─────▲───┘  │     │ ┌───────────────────┐ │
+                   │      │     │ │                   │ │    creates & owns
+        watches &  └──────┴─────┼─┤ Module controller ├─┼────────────────────┐
+        updates status          │ │                   │ │                    │
+                                │ └───────────────────┘ │           ┌────────▼────────┐
+                                │                       │           │                 │
+                                │                       │           │ NodeModuleState │
+┌────────────┐                  │ ┌─────────────────┐   │           │                 │
+│            │                  │ │                 │   │           └────────▲────────┘
+│ KMM worker ◄──────────────────┼─┤ NodeModuleState │   │                    │
+│    Pod     │  creates & owns  │ │ controller      ├───┼────────────────────┘
+│            │                  │ │                 │   │   watches &
+└──────┬─────┘                  │ └─────────────────┘   │   updates status
+       │                        │                       │
+       │                        │          KMM operator │
+       │                        └───────────────────────┘
+       │
+       │
+       │                        ┌──────────────────────┐
+       │                        │                      │
+       └────────────────────────► registry.example.org │
+                pulls from      │                      │
+                                └──────────────────────┘
+```
+
+#### The operator
+
+The operator runs two main controllers: the `Module` controller and the `NodeModuleState` controller.
+
+The `Module` controller watches both `Module` and `Node` resources to determine which kmods should be loaded on which
+`Node`.
+It creates one `NodeModuleState` resource per `Node` and `Module` that targets that node.  
```

Comment:
I thought that we wanted to have a single `NodeModuleState` for the states of all modules on a nodes that has a least one module.

What is the benefit of having a `NodeModuleState` per module per node VS having a single `NodeModule*s*State` per node for all modules?